File: Java/32. Longest Valid Parentheses.java
class Solution {
    public int longestValidParentheses(String s) {
        int n = s.length();
        int res = 0;
        int l = -1;
        Stack<Integer> stk = new Stack<Integer>();
        for(int i = 0; i < n; i++){
            if(s.charAt(i) == '('){
                stk.push(i);
            }
            else{
                if(stk.isEmpty()){
                    l = i;
                }
                else{
                    stk.pop();
                    if(stk.isEmpty()){
                        res=Math.max(res,i-l);
                    }
                    else{
                        res=Math.max(res,i-stk.peek());
                    }
                }
            }
        }
        return res;
    }
}

File: C#/32. Longest Valid Parentheses.cs
public class Solution {
    public int LongestValidParentheses(string s) {
        int n = s.Length;
        int res = 0;
        int l = -1;
        Stack<int> stk = new Stack<int>();
        for(int i = 0; i < n; i++){
            if(s[i] == '('){
                stk.Push(i);
            }
            else{
                if(stk.Count==0){
                    l = i;
                }
                else{
                    stk.Pop();
                    if(stk.Count==0){
                        res=Math.Max(res,i-l);
                    }
                    else{
                        res=Math.Max(res,i-stk.Peek());
                    }
                }
            }
        }
        return res;
    }
}
